diff --git a/Makefile b/Makefile
index f8c820e..9668c93 100644
--- a/Makefile
+++ b/Makefile
@@ -139,6 +139,9 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_history\
+	$U/_dummyproc\
+	$U/_testprocinfo\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
diff --git a/kernel/defs.h b/kernel/defs.h
index d1b6bb9..8c321b1 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -8,6 +8,9 @@ struct spinlock;
 struct sleeplock;
 struct stat;
 struct superblock;
+struct pstat;
+
+
 
 // bio.c
 void            binit(void);
@@ -187,3 +190,10 @@ void            virtio_disk_intr(void);
 
 // number of elements in fixed-size array
 #define NELEM(x) (sizeof(x)/sizeof((x)[0]))
+
+
+
+int             settickets(int);
+int             getpinfo(struct pstat*);
+void            syscall_stats_init(void); 
+void            priority_boosting(void);
\ No newline at end of file
diff --git a/kernel/main.c b/kernel/main.c
index f0d3171..acdec11 100644
--- a/kernel/main.c
+++ b/kernel/main.c
@@ -4,6 +4,7 @@
 #include "riscv.h"
 #include "defs.h"
 
+
 volatile static int started = 0;
 
 // start() jumps here in supervisor mode on all CPUs.
@@ -20,6 +21,7 @@ main()
     kvminit();       // create kernel page table
     kvminithart();   // turn on paging
     procinit();      // process table
+    syscall_stats_init();
     trapinit();      // trap vectors
     trapinithart();  // install kernel trap vector
     plicinit();      // set up interrupt controller
diff --git a/kernel/param.h b/kernel/param.h
index 80ec6d3..cd3c179 100644
--- a/kernel/param.h
+++ b/kernel/param.h
@@ -12,4 +12,8 @@
 #define FSSIZE       2000  // size of file system in blocks
 #define MAXPATH      128   // maximum file path name
 #define USERSTACK    1     // user stack pages
-
+#define TIME_LIMIT_1 1
+#define TIME_LIMIT_2 2
+#define BOOST_INTERVAL 64
+#define DEFAULT_TICKET_COUNT 10
+#define PRINT_SCHEDULING 0
diff --git a/kernel/proc.c b/kernel/proc.c
index 130d9ce..5463aa9 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -5,6 +5,7 @@
 #include "spinlock.h"
 #include "proc.h"
 #include "defs.h"
+#include "pstat.h"
 
 struct cpu cpus[NCPU];
 
@@ -29,40 +30,39 @@ struct spinlock wait_lock;
 // Allocate a page for each process's kernel stack.
 // Map it high in memory, followed by an invalid
 // guard page.
-void
-proc_mapstacks(pagetable_t kpgtbl)
+void proc_mapstacks(pagetable_t kpgtbl)
 {
   struct proc *p;
-  
-  for(p = proc; p < &proc[NPROC]; p++) {
+
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
     char *pa = kalloc();
-    if(pa == 0)
+    if (pa == 0)
       panic("kalloc");
-    uint64 va = KSTACK((int) (p - proc));
+    uint64 va = KSTACK((int)(p - proc));
     kvmmap(kpgtbl, va, (uint64)pa, PGSIZE, PTE_R | PTE_W);
   }
 }
 
 // initialize the proc table.
-void
-procinit(void)
+void procinit(void)
 {
   struct proc *p;
-  
+
   initlock(&pid_lock, "nextpid");
   initlock(&wait_lock, "wait_lock");
-  for(p = proc; p < &proc[NPROC]; p++) {
-      initlock(&p->lock, "proc");
-      p->state = UNUSED;
-      p->kstack = KSTACK((int) (p - proc));
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
+    initlock(&p->lock, "proc");
+    p->state = UNUSED;
+    p->kstack = KSTACK((int)(p - proc));
   }
 }
 
 // Must be called with interrupts disabled,
 // to prevent race with process being moved
 // to a different CPU.
-int
-cpuid()
+int cpuid()
 {
   int id = r_tp();
   return id;
@@ -70,7 +70,7 @@ cpuid()
 
 // Return this CPU's cpu struct.
 // Interrupts must be disabled.
-struct cpu*
+struct cpu *
 mycpu(void)
 {
   int id = cpuid();
@@ -79,7 +79,7 @@ mycpu(void)
 }
 
 // Return the current struct proc *, or zero if none.
-struct proc*
+struct proc *
 myproc(void)
 {
   push_off();
@@ -89,11 +89,10 @@ myproc(void)
   return p;
 }
 
-int
-allocpid()
+int allocpid()
 {
   int pid;
-  
+
   acquire(&pid_lock);
   pid = nextpid;
   nextpid = nextpid + 1;
@@ -106,16 +105,20 @@ allocpid()
 // If found, initialize state required to run in the kernel,
 // and return with p->lock held.
 // If there are no free procs, or a memory allocation fails, return 0.
-static struct proc*
+static struct proc *
 allocproc(void)
 {
   struct proc *p;
 
-  for(p = proc; p < &proc[NPROC]; p++) {
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
     acquire(&p->lock);
-    if(p->state == UNUSED) {
+    if (p->state == UNUSED)
+    {
       goto found;
-    } else {
+    }
+    else
+    {
       release(&p->lock);
     }
   }
@@ -126,7 +129,8 @@ found:
   p->state = USED;
 
   // Allocate a trapframe page.
-  if((p->trapframe = (struct trapframe *)kalloc()) == 0){
+  if ((p->trapframe = (struct trapframe *)kalloc()) == 0)
+  {
     freeproc(p);
     release(&p->lock);
     return 0;
@@ -134,7 +138,8 @@ found:
 
   // An empty user page table.
   p->pagetable = proc_pagetable(p);
-  if(p->pagetable == 0){
+  if (p->pagetable == 0)
+  {
     freeproc(p);
     release(&p->lock);
     return 0;
@@ -146,6 +151,14 @@ found:
   p->context.ra = (uint64)forkret;
   p->context.sp = p->kstack + PGSIZE;
 
+  p->queue = 1;
+  p->num_original_tickets = DEFAULT_TICKET_COUNT;
+  p->num_curr_tickets = DEFAULT_TICKET_COUNT;
+  p->total_time_slices = 0;
+  p->current_time_slices = 0;
+
+  // printf("%s %d\n",p->name,p->pid);
+
   return p;
 }
 
@@ -155,10 +168,10 @@ found:
 static void
 freeproc(struct proc *p)
 {
-  if(p->trapframe)
-    kfree((void*)p->trapframe);
+  if (p->trapframe)
+    kfree((void *)p->trapframe);
   p->trapframe = 0;
-  if(p->pagetable)
+  if (p->pagetable)
     proc_freepagetable(p->pagetable, p->sz);
   p->pagetable = 0;
   p->sz = 0;
@@ -180,23 +193,25 @@ proc_pagetable(struct proc *p)
 
   // An empty page table.
   pagetable = uvmcreate();
-  if(pagetable == 0)
+  if (pagetable == 0)
     return 0;
 
   // map the trampoline code (for system call return)
   // at the highest user virtual address.
   // only the supervisor uses it, on the way
   // to/from user space, so not PTE_U.
-  if(mappages(pagetable, TRAMPOLINE, PGSIZE,
-              (uint64)trampoline, PTE_R | PTE_X) < 0){
+  if (mappages(pagetable, TRAMPOLINE, PGSIZE,
+               (uint64)trampoline, PTE_R | PTE_X) < 0)
+  {
     uvmfree(pagetable, 0);
     return 0;
   }
 
   // map the trapframe page just below the trampoline page, for
   // trampoline.S.
-  if(mappages(pagetable, TRAPFRAME, PGSIZE,
-              (uint64)(p->trapframe), PTE_R | PTE_W) < 0){
+  if (mappages(pagetable, TRAPFRAME, PGSIZE,
+               (uint64)(p->trapframe), PTE_R | PTE_W) < 0)
+  {
     uvmunmap(pagetable, TRAMPOLINE, 1, 0);
     uvmfree(pagetable, 0);
     return 0;
@@ -207,8 +222,7 @@ proc_pagetable(struct proc *p)
 
 // Free a process's page table, and free the
 // physical memory it refers to.
-void
-proc_freepagetable(pagetable_t pagetable, uint64 sz)
+void proc_freepagetable(pagetable_t pagetable, uint64 sz)
 {
   uvmunmap(pagetable, TRAMPOLINE, 1, 0);
   uvmunmap(pagetable, TRAPFRAME, 1, 0);
@@ -219,32 +233,30 @@ proc_freepagetable(pagetable_t pagetable, uint64 sz)
 // assembled from ../user/initcode.S
 // od -t xC ../user/initcode
 uchar initcode[] = {
-  0x17, 0x05, 0x00, 0x00, 0x13, 0x05, 0x45, 0x02,
-  0x97, 0x05, 0x00, 0x00, 0x93, 0x85, 0x35, 0x02,
-  0x93, 0x08, 0x70, 0x00, 0x73, 0x00, 0x00, 0x00,
-  0x93, 0x08, 0x20, 0x00, 0x73, 0x00, 0x00, 0x00,
-  0xef, 0xf0, 0x9f, 0xff, 0x2f, 0x69, 0x6e, 0x69,
-  0x74, 0x00, 0x00, 0x24, 0x00, 0x00, 0x00, 0x00,
-  0x00, 0x00, 0x00, 0x00
-};
+    0x17, 0x05, 0x00, 0x00, 0x13, 0x05, 0x45, 0x02,
+    0x97, 0x05, 0x00, 0x00, 0x93, 0x85, 0x35, 0x02,
+    0x93, 0x08, 0x70, 0x00, 0x73, 0x00, 0x00, 0x00,
+    0x93, 0x08, 0x20, 0x00, 0x73, 0x00, 0x00, 0x00,
+    0xef, 0xf0, 0x9f, 0xff, 0x2f, 0x69, 0x6e, 0x69,
+    0x74, 0x00, 0x00, 0x24, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00};
 
 // Set up first user process.
-void
-userinit(void)
+void userinit(void)
 {
   struct proc *p;
 
   p = allocproc();
   initproc = p;
-  
+
   // allocate one user page and copy initcode's instructions
   // and data into it.
   uvmfirst(p->pagetable, initcode, sizeof(initcode));
   p->sz = PGSIZE;
 
   // prepare for the very first "return" from kernel to user.
-  p->trapframe->epc = 0;      // user program counter
-  p->trapframe->sp = PGSIZE;  // user stack pointer
+  p->trapframe->epc = 0;     // user program counter
+  p->trapframe->sp = PGSIZE; // user stack pointer
 
   safestrcpy(p->name, "initcode", sizeof(p->name));
   p->cwd = namei("/");
@@ -256,18 +268,21 @@ userinit(void)
 
 // Grow or shrink user memory by n bytes.
 // Return 0 on success, -1 on failure.
-int
-growproc(int n)
+int growproc(int n)
 {
   uint64 sz;
   struct proc *p = myproc();
 
   sz = p->sz;
-  if(n > 0){
-    if((sz = uvmalloc(p->pagetable, sz, sz + n, PTE_W)) == 0) {
+  if (n > 0)
+  {
+    if ((sz = uvmalloc(p->pagetable, sz, sz + n, PTE_W)) == 0)
+    {
       return -1;
     }
-  } else if(n < 0){
+  }
+  else if (n < 0)
+  {
     sz = uvmdealloc(p->pagetable, sz, sz + n);
   }
   p->sz = sz;
@@ -276,20 +291,21 @@ growproc(int n)
 
 // Create a new process, copying the parent.
 // Sets up child kernel stack to return as if from fork() system call.
-int
-fork(void)
+int fork(void)
 {
   int i, pid;
   struct proc *np;
   struct proc *p = myproc();
 
   // Allocate process.
-  if((np = allocproc()) == 0){
+  if ((np = allocproc()) == 0)
+  {
     return -1;
   }
 
   // Copy user memory from parent to child.
-  if(uvmcopy(p->pagetable, np->pagetable, p->sz) < 0){
+  if (uvmcopy(p->pagetable, np->pagetable, p->sz) < 0)
+  {
     freeproc(np);
     release(&np->lock);
     return -1;
@@ -303,8 +319,8 @@ fork(void)
   np->trapframe->a0 = 0;
 
   // increment reference counts on open file descriptors.
-  for(i = 0; i < NOFILE; i++)
-    if(p->ofile[i])
+  for (i = 0; i < NOFILE; i++)
+    if (p->ofile[i])
       np->ofile[i] = filedup(p->ofile[i]);
   np->cwd = idup(p->cwd);
 
@@ -312,6 +328,12 @@ fork(void)
 
   pid = np->pid;
 
+  np->num_original_tickets = p->num_original_tickets;
+  np->num_curr_tickets = p->num_original_tickets;
+  np->queue = 1;
+  np->total_time_slices = 0;
+  np->current_time_slices = 0;
+
   release(&np->lock);
 
   acquire(&wait_lock);
@@ -327,13 +349,14 @@ fork(void)
 
 // Pass p's abandoned children to init.
 // Caller must hold wait_lock.
-void
-reparent(struct proc *p)
+void reparent(struct proc *p)
 {
   struct proc *pp;
 
-  for(pp = proc; pp < &proc[NPROC]; pp++){
-    if(pp->parent == p){
+  for (pp = proc; pp < &proc[NPROC]; pp++)
+  {
+    if (pp->parent == p)
+    {
       pp->parent = initproc;
       wakeup(initproc);
     }
@@ -343,17 +366,18 @@ reparent(struct proc *p)
 // Exit the current process.  Does not return.
 // An exited process remains in the zombie state
 // until its parent calls wait().
-void
-exit(int status)
+void exit(int status)
 {
   struct proc *p = myproc();
 
-  if(p == initproc)
+  if (p == initproc)
     panic("init exiting");
 
   // Close all open files.
-  for(int fd = 0; fd < NOFILE; fd++){
-    if(p->ofile[fd]){
+  for (int fd = 0; fd < NOFILE; fd++)
+  {
+    if (p->ofile[fd])
+    {
       struct file *f = p->ofile[fd];
       fileclose(f);
       p->ofile[fd] = 0;
@@ -372,7 +396,7 @@ exit(int status)
 
   // Parent might be sleeping in wait().
   wakeup(p->parent);
-  
+
   acquire(&p->lock);
 
   p->xstate = status;
@@ -387,8 +411,7 @@ exit(int status)
 
 // Wait for a child process to exit and return its pid.
 // Return -1 if this process has no children.
-int
-wait(uint64 addr)
+int wait(uint64 addr)
 {
   struct proc *pp;
   int havekids, pid;
@@ -396,20 +419,25 @@ wait(uint64 addr)
 
   acquire(&wait_lock);
 
-  for(;;){
+  for (;;)
+  {
     // Scan through table looking for exited children.
     havekids = 0;
-    for(pp = proc; pp < &proc[NPROC]; pp++){
-      if(pp->parent == p){
+    for (pp = proc; pp < &proc[NPROC]; pp++)
+    {
+      if (pp->parent == p)
+      {
         // make sure the child isn't still in exit() or swtch().
         acquire(&pp->lock);
 
         havekids = 1;
-        if(pp->state == ZOMBIE){
+        if (pp->state == ZOMBIE)
+        {
           // Found one.
           pid = pp->pid;
-          if(addr != 0 && copyout(p->pagetable, addr, (char *)&pp->xstate,
-                                  sizeof(pp->xstate)) < 0) {
+          if (addr != 0 && copyout(p->pagetable, addr, (char *)&pp->xstate,
+                                   sizeof(pp->xstate)) < 0)
+          {
             release(&pp->lock);
             release(&wait_lock);
             return -1;
@@ -424,13 +452,14 @@ wait(uint64 addr)
     }
 
     // No point waiting if we don't have any children.
-    if(!havekids || killed(p)){
+    if (!havekids || killed(p))
+    {
       release(&wait_lock);
       return -1;
     }
-    
+
     // Wait for a child to exit.
-    sleep(p, &wait_lock);  //DOC: wait-sleep
+    sleep(p, &wait_lock); // DOC: wait-sleep
   }
 }
 
@@ -441,38 +470,218 @@ wait(uint64 addr)
 //  - swtch to start running that process.
 //  - eventually that process transfers control
 //    via swtch back to the scheduler.
-void
-scheduler(void)
+
+static unsigned int rand_state = 123456789;
+
+// https://en.wikipedia.org/wiki/Xorshift#xorshift*
+
+static unsigned int xorshift32(void)
+{
+  unsigned int x = rand_state;
+  x ^= x >> 12;
+  x ^= x << 25;
+  x ^= x >> 27;
+  rand_state = x;
+  return x;
+}
+
+// Returns pseudo-random number in [1, n]
+int randrange(int n)
+{
+  unsigned int r = xorshift32();
+  return (r % n) + 1;
+}
+
+void priority_boosting(void)
+{
+  int runable = 0;
+  struct proc *p;
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
+    acquire(&p->lock);
+    if (p->state == RUNNABLE)
+    {
+      p->queue = 1;
+      runable++;
+    }
+    release(&p->lock);
+  }
+
+  if (PRINT_SCHEDULING && runable)
+    printf("Priority boosting at tick %d for %d runnable process\n\n", ticks, runable);
+}
+
+void scheduler(void)
 {
   struct proc *p;
   struct cpu *c = mycpu();
 
   c->proc = 0;
-  for(;;){
+  for (;;)
+  {
     // The most recent process to run may have had interrupts
     // turned off; enable them to avoid a deadlock if all
     // processes are waiting.
     intr_on();
 
+    int all_tickets_zero = 1;
+    int num_runable = 0;
+    for (p = proc; p < &proc[NPROC]; p++)
+    {
+      acquire(&p->lock);
+      if (p->state == RUNNABLE)
+      {
+        num_runable++;
+        if (p->num_curr_tickets > 0)
+        {
+          all_tickets_zero = 0;
+          release(&p->lock);
+          break;
+        }
+      }
+      release(&p->lock);
+    }
+    if (num_runable != 0 && all_tickets_zero)
+    {
+      if (PRINT_SCHEDULING)
+        printf("Reinitializing tickets at tick %d\n\n", ticks);
+
+      for (p = proc; p < &proc[NPROC]; p++)
+      {
+        acquire(&p->lock);
+        if (p->state == RUNNABLE)
+        {
+          p->num_curr_tickets = p->num_original_tickets;
+        }
+        release(&p->lock);
+      }
+    }
+
     int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
+
+    struct spinlock total_tickets_lock = {0};
+    initlock(&total_tickets_lock, "total_tickets_lock");
+
+    int total_tickets = 0;
+    acquire(&total_tickets_lock);
+    for (p = proc; p < &proc[NPROC]; p++)
+    {
       acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
-        found = 1;
+      if (p->state == RUNNABLE && p->queue == 1 && p->num_curr_tickets > 0)
+      {
+        total_tickets += p->num_curr_tickets;
       }
       release(&p->lock);
     }
-    if(found == 0) {
+    release(&total_tickets_lock);
+
+  
+
+    if (total_tickets > 0)
+    {
+      int winner = randrange(total_tickets);
+      int current = 0;
+      for (p = proc; p < &proc[NPROC]; p++)
+      {
+        acquire(&p->lock);
+        if (p->state == RUNNABLE && p->queue == 1 && p->num_curr_tickets > 0)
+        {
+          current += p->num_curr_tickets;
+          if (current >= winner)
+          {
+            while (p->state == RUNNABLE && p->current_time_slices < TIME_LIMIT_1)
+            {
+              p->state = RUNNING;
+              c->proc = p;
+              found = 1;
+
+              swtch(&c->context, &p->context);
+              int id = cpuid();
+              if (PRINT_SCHEDULING)
+                printf("[CPU %d][Tick %d] SCHEDULING: PID %d, queue=%d, curr_tickets=%d/%d, curr_slices=%d/%d, time_slices=%d\n\n",
+                       id, ticks, p->pid, p->queue, p->num_curr_tickets, p->num_original_tickets,
+                       p->current_time_slices, (p->queue == 1 ? TIME_LIMIT_1 : TIME_LIMIT_2), p->total_time_slices);
+
+              // Process is done running for now.
+              // It should have changed its p->state before coming back.
+              c->proc = 0;
+              p->current_time_slices++;
+              p->total_time_slices++;
+              if (p->num_curr_tickets)
+                p->num_curr_tickets--;
+            }
+
+            if (p->current_time_slices >= TIME_LIMIT_1)
+            {
+
+              int id = cpuid();
+              if (PRINT_SCHEDULING)
+                printf("[CPU %d][Tick %d] DEMOTING: PID %d, queue=%d, curr_tickets=%d/%d, curr_slices=%d/%d, time_slices=%d to QUEUE 2\n\n",
+                  id, ticks, p->pid, p->queue, p->num_curr_tickets, p->num_original_tickets,
+                  p->current_time_slices, (p->queue == 1 ? TIME_LIMIT_1 : TIME_LIMIT_2), p->total_time_slices);
+
+              p->queue = 2;
+            }
+
+            p->current_time_slices = 0;
+          }
+          release(&p->lock);
+          break;
+        }
+        release(&p->lock);
+      }
+    }
+
+    if (!found)
+    {
+      for (p = proc; p < &proc[NPROC]; p++)
+      {
+        acquire(&p->lock);
+        if (p->queue == 2 && p->state == RUNNABLE)
+        {
+          // Switch to chosen process.  It is the process's job
+          // to release its lock and then reacquire it
+          // before jumping back to us.
+          while (p->state == RUNNABLE && p->current_time_slices< TIME_LIMIT_2)
+          {
+            p->state = RUNNING;
+            c->proc = p;
+            found = 1;
+            swtch(&c->context, &p->context);
+            int id = cpuid();
+            if (PRINT_SCHEDULING)
+              printf("[CPU %d][Tick %d] SCHEDULING: PID %d, queue=%d, curr_tickets=%d/%d, curr_slices=%d/%d, time_slices=%d\n\n",
+                id, ticks, p->pid, p->queue, p->num_curr_tickets, p->num_original_tickets,
+                p->current_time_slices, (p->queue == 1 ? TIME_LIMIT_1 : TIME_LIMIT_2), p->total_time_slices);
+            
+                    
+
+            // Process is done running for now.
+            // It should have changed its p->state before coming back.
+            c->proc = 0;
+            p->current_time_slices++;
+            p->total_time_slices++;
+           
+          }
+
+          if (p->current_time_slices < TIME_LIMIT_2)
+          {
+
+            int id = cpuid();
+            if (PRINT_SCHEDULING)
+              printf("[CPU %d][Tick %d] PROMOTING: PID %d, queue=%d, curr_tickets=%d/%d, curr_slices=%d/%d, time_slices=%d to QUEUE 1\n\n",
+                     id, ticks, p->pid, p->queue, p->num_curr_tickets, p->num_original_tickets,
+                     p->current_time_slices, (p->queue == 1 ? TIME_LIMIT_1 : TIME_LIMIT_2), p->total_time_slices);
+
+            p->queue = 1;
+          }
+          p->current_time_slices = 0;
+        }
+        release(&p->lock);
+      }
+    }
+    if (found == 0)
+    {
       // nothing to run; stop running on this core until an interrupt.
       intr_on();
       asm volatile("wfi");
@@ -487,19 +696,18 @@ scheduler(void)
 // be proc->intena and proc->noff, but that would
 // break in the few places where a lock is held but
 // there's no process.
-void
-sched(void)
+void sched(void)
 {
   int intena;
   struct proc *p = myproc();
 
-  if(!holding(&p->lock))
+  if (!holding(&p->lock))
     panic("sched p->lock");
-  if(mycpu()->noff != 1)
+  if (mycpu()->noff != 1)
     panic("sched locks");
-  if(p->state == RUNNING)
+  if (p->state == RUNNING)
     panic("sched running");
-  if(intr_get())
+  if (intr_get())
     panic("sched interruptible");
 
   intena = mycpu()->intena;
@@ -508,8 +716,7 @@ sched(void)
 }
 
 // Give up the CPU for one scheduling round.
-void
-yield(void)
+void yield(void)
 {
   struct proc *p = myproc();
   acquire(&p->lock);
@@ -520,15 +727,15 @@ yield(void)
 
 // A fork child's very first scheduling by scheduler()
 // will swtch to forkret.
-void
-forkret(void)
+void forkret(void)
 {
   static int first = 1;
 
   // Still holding p->lock from scheduler.
   release(&myproc()->lock);
 
-  if (first) {
+  if (first)
+  {
     // File system initialization must be run in the context of a
     // regular process (e.g., because it calls sleep), and thus cannot
     // be run from main().
@@ -544,11 +751,10 @@ forkret(void)
 
 // Atomically release lock and sleep on chan.
 // Reacquires lock when awakened.
-void
-sleep(void *chan, struct spinlock *lk)
+void sleep(void *chan, struct spinlock *lk)
 {
   struct proc *p = myproc();
-  
+
   // Must acquire p->lock in order to
   // change p->state and then call sched.
   // Once we hold p->lock, we can be
@@ -556,7 +762,7 @@ sleep(void *chan, struct spinlock *lk)
   // (wakeup locks p->lock),
   // so it's okay to release lk.
 
-  acquire(&p->lock);  //DOC: sleeplock1
+  acquire(&p->lock); // DOC: sleeplock1
   release(lk);
 
   // Go to sleep.
@@ -575,15 +781,17 @@ sleep(void *chan, struct spinlock *lk)
 
 // Wake up all processes sleeping on chan.
 // Must be called without any p->lock.
-void
-wakeup(void *chan)
+void wakeup(void *chan)
 {
   struct proc *p;
 
-  for(p = proc; p < &proc[NPROC]; p++) {
-    if(p != myproc()){
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
+    if (p != myproc())
+    {
       acquire(&p->lock);
-      if(p->state == SLEEPING && p->chan == chan) {
+      if (p->state == SLEEPING && p->chan == chan)
+      {
         p->state = RUNNABLE;
       }
       release(&p->lock);
@@ -594,16 +802,18 @@ wakeup(void *chan)
 // Kill the process with the given pid.
 // The victim won't exit until it tries to return
 // to user space (see usertrap() in trap.c).
-int
-kill(int pid)
+int kill(int pid)
 {
   struct proc *p;
 
-  for(p = proc; p < &proc[NPROC]; p++){
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
     acquire(&p->lock);
-    if(p->pid == pid){
+    if (p->pid == pid)
+    {
       p->killed = 1;
-      if(p->state == SLEEPING){
+      if (p->state == SLEEPING)
+      {
         // Wake process from sleep().
         p->state = RUNNABLE;
       }
@@ -615,19 +825,17 @@ kill(int pid)
   return -1;
 }
 
-void
-setkilled(struct proc *p)
+void setkilled(struct proc *p)
 {
   acquire(&p->lock);
   p->killed = 1;
   release(&p->lock);
 }
 
-int
-killed(struct proc *p)
+int killed(struct proc *p)
 {
   int k;
-  
+
   acquire(&p->lock);
   k = p->killed;
   release(&p->lock);
@@ -637,13 +845,15 @@ killed(struct proc *p)
 // Copy to either a user address, or kernel address,
 // depending on usr_dst.
 // Returns 0 on success, -1 on error.
-int
-either_copyout(int user_dst, uint64 dst, void *src, uint64 len)
+int either_copyout(int user_dst, uint64 dst, void *src, uint64 len)
 {
   struct proc *p = myproc();
-  if(user_dst){
+  if (user_dst)
+  {
     return copyout(p->pagetable, dst, src, len);
-  } else {
+  }
+  else
+  {
     memmove((char *)dst, src, len);
     return 0;
   }
@@ -652,14 +862,16 @@ either_copyout(int user_dst, uint64 dst, void *src, uint64 len)
 // Copy from either a user address, or kernel address,
 // depending on usr_src.
 // Returns 0 on success, -1 on error.
-int
-either_copyin(void *dst, int user_src, uint64 src, uint64 len)
+int either_copyin(void *dst, int user_src, uint64 src, uint64 len)
 {
   struct proc *p = myproc();
-  if(user_src){
+  if (user_src)
+  {
     return copyin(p->pagetable, dst, src, len);
-  } else {
-    memmove(dst, (char*)src, len);
+  }
+  else
+  {
+    memmove(dst, (char *)src, len);
     return 0;
   }
 }
@@ -667,25 +879,24 @@ either_copyin(void *dst, int user_src, uint64 src, uint64 len)
 // Print a process listing to console.  For debugging.
 // Runs when user types ^P on console.
 // No lock to avoid wedging a stuck machine further.
-void
-procdump(void)
+void procdump(void)
 {
   static char *states[] = {
-  [UNUSED]    "unused",
-  [USED]      "used",
-  [SLEEPING]  "sleep ",
-  [RUNNABLE]  "runble",
-  [RUNNING]   "run   ",
-  [ZOMBIE]    "zombie"
-  };
+      [UNUSED] "unused",
+      [USED] "used",
+      [SLEEPING] "sleep ",
+      [RUNNABLE] "runble",
+      [RUNNING] "run   ",
+      [ZOMBIE] "zombie"};
   struct proc *p;
   char *state;
 
   printf("\n");
-  for(p = proc; p < &proc[NPROC]; p++){
-    if(p->state == UNUSED)
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
+    if (p->state == UNUSED)
       continue;
-    if(p->state >= 0 && p->state < NELEM(states) && states[p->state])
+    if (p->state >= 0 && p->state < NELEM(states) && states[p->state])
       state = states[p->state];
     else
       state = "???";
@@ -693,3 +904,48 @@ procdump(void)
     printf("\n");
   }
 }
+
+int settickets(int n)
+{
+  struct proc *p = myproc();
+
+  if (n < 1)
+  {
+    n = DEFAULT_TICKET_COUNT;
+  }
+
+  acquire(&p->lock);
+  p->num_original_tickets = n;
+  p->num_curr_tickets = n;
+  release(&p->lock);
+
+  return 0;
+}
+
+int getpinfo(struct pstat *ps)
+{
+  struct proc *p;
+  for (int i = 0; i < NPROC; i++)
+  {
+    ps->pid[i] = 0;
+    ps->inuse[i] = 0;
+    ps->inQ[i] = 0;
+    ps->tickets_original[i] = 0;
+    ps->tickets_current[i] = 0;
+    ps->time_slices[i] = 0;
+  }
+
+  for (p = proc; p < &proc[NPROC]; p++)
+  {
+    acquire(&p->lock);
+    int i = p - proc;
+    ps->pid[i] = p->pid;
+    ps->inuse[i] = (p->state == RUNNABLE || p->state == RUNNING);
+    ps->inQ[i] = p->queue;
+    ps->tickets_original[i] = p->num_original_tickets;
+    ps->tickets_current[i] = p->num_curr_tickets;
+    ps->time_slices[i] = p->total_time_slices;
+    release(&p->lock);
+  }
+  return 0;
+}
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..3ff9548 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -83,6 +83,13 @@ enum procstate { UNUSED, USED, SLEEPING, RUNNABLE, RUNNING, ZOMBIE };
 
 // Per-process state
 struct proc {
+
+  
+    int queue;                   
+    int num_original_tickets;          
+    int num_curr_tickets;            
+    int total_time_slices;           
+    int current_time_slices;            
   struct spinlock lock;
 
   // p->lock must be held when using these:
diff --git a/kernel/pstat.h b/kernel/pstat.h
new file mode 100644
index 0000000..a7de853
--- /dev/null
+++ b/kernel/pstat.h
@@ -0,0 +1,12 @@
+#ifndef _PSTAT_H_
+#define _PSTAT_H_
+#include "param.h"
+struct pstat {
+    int pid[NPROC]; // the process ID of each process
+    int inuse[NPROC]; // whether this slot of the process table is being used (1 or 0 )
+    int inQ[NPROC]; // which queue the process is currently in
+    int tickets_original[NPROC]; // the number of tickets each process originally had
+    int tickets_current[NPROC]; // the number of tickets each process currently has
+    int time_slices[NPROC]; // the number of time slices each process has been scheduled
+};
+#endif // _PSTAT_H_
\ No newline at end of file
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..34272df 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -6,6 +6,29 @@
 #include "proc.h"
 #include "syscall.h"
 #include "defs.h"
+#include "syscall_stat.h"
+#include "pstat.h"
+
+
+struct syscall_stat syscall_stats[NSYSCALLS+1];
+struct spinlock syscall_locks[NSYSCALLS+1];
+
+
+char *syscall_names[NSYSCALLS+1] = {
+  "","fork", "exit", "wait", "pipe", "read", "kill", "exec", "fstat", "chdir",
+  "dup", "getpid", "sbrk", "sleep", "uptime", "open", "write", "mknod",
+  "unlink", "link", "mkdir", "close", "history","settickets","getpinfo"
+};
+
+void
+syscall_stats_init(void) {
+  for(int i = 1; i <= NSYSCALLS; i++) {
+    safestrcpy(syscall_stats[i].syscall_name, syscall_names[i], 16);
+    syscall_stats[i].count = 0;
+    syscall_stats[i].accum_time = 0;
+    initlock(&syscall_locks[i], "syscall_lock");
+  }
+}
 
 // Fetch the uint64 at addr from the current process.
 int
@@ -101,6 +124,9 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_history(void);
+extern uint64 sys_settickets(void);
+extern uint64 sys_getpinfo(void);
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,8 +152,38 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_history] sys_history,
+[SYS_settickets] sys_settickets,
+[SYS_getpinfo]   sys_getpinfo,
 };
 
+int
+history(int sysno, struct syscall_stat *userptr)
+{
+
+  struct proc *p = myproc();
+  acquire(&syscall_locks[sysno]);
+  int result = copyout(p->pagetable, (uint64)userptr, (char *)&syscall_stats[sysno], sizeof(struct syscall_stat));
+  release(&syscall_locks[sysno]);
+  return result;
+}
+
+uint64
+sys_history(void)
+{
+  int sysno;
+  uint64 userptr;
+  argint(0, &sysno);
+  if (sysno < 1 || sysno > NSYSCALLS) {
+    return -1; 
+  }
+  argaddr(1, &userptr);
+  
+  return history(sysno, (struct syscall_stat *)userptr);
+}
+
+
+
 void
 syscall(void)
 {
@@ -138,10 +194,18 @@ syscall(void)
   if(num > 0 && num < NELEM(syscalls) && syscalls[num]) {
     // Use num to lookup the system call function for num, call it,
     // and store its return value in p->trapframe->a0
-    p->trapframe->a0 = syscalls[num]();
+      int start_ticks = ticks;
+      p->trapframe->a0 = syscalls[num]();
+      int end_ticks = ticks;
+      acquire(&syscall_locks[num]);
+      syscall_stats[num].count++;
+      syscall_stats[num].accum_time += (end_ticks - start_ticks);
+      release(&syscall_locks[num]);
   } else {
     printf("%d %s: unknown sys call %d\n",
             p->pid, p->name, num);
     p->trapframe->a0 = -1;
   }
 }
+
+
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..75d1dc2 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,6 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_history 22
+#define SYS_settickets 23
+#define SYS_getpinfo   24
diff --git a/kernel/syscall_stat.h b/kernel/syscall_stat.h
new file mode 100644
index 0000000..dd26698
--- /dev/null
+++ b/kernel/syscall_stat.h
@@ -0,0 +1,7 @@
+#define NSYSCALLS 24
+
+struct syscall_stat {
+    char syscall_name[16];
+    int count;
+    int accum_time;
+  };
\ No newline at end of file
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 3b4d5bd..5cbac8e 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -5,6 +5,7 @@
 #include "memlayout.h"
 #include "spinlock.h"
 #include "proc.h"
+#include "pstat.h"
 
 uint64
 sys_exit(void)
@@ -91,3 +92,33 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+
+
+uint64
+sys_settickets(void) {
+    int n;
+    argint(0, &n);
+    return settickets(n);
+}
+
+
+uint64
+sys_getpinfo(void)
+{
+    uint64 addr;
+    struct pstat ps;
+    struct proc *p = myproc();
+    argaddr(0, &addr) ;
+
+    if (addr < 0 || addr == 0)
+        return -1;
+
+    if (getpinfo(&ps) < 0)
+        return -1;
+
+    if (copyout(p->pagetable, addr, (char *)&ps, sizeof(struct pstat)) < 0)
+        return -1;
+
+    return 0;
+}
\ No newline at end of file
diff --git a/kernel/trap.c b/kernel/trap.c
index d454a7d..b445623 100644
--- a/kernel/trap.c
+++ b/kernel/trap.c
@@ -166,6 +166,10 @@ clockintr()
   if(cpuid() == 0){
     acquire(&tickslock);
     ticks++;
+    if(ticks%BOOST_INTERVAL==0)
+    {
+      priority_boosting();
+    }
     wakeup(&ticks);
     release(&tickslock);
   }
diff --git a/user/dummyproc.c b/user/dummyproc.c
new file mode 100644
index 0000000..75613a1
--- /dev/null
+++ b/user/dummyproc.c
@@ -0,0 +1,47 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/pstat.h"
+#define NUM_ITERATION 1000000000
+
+int
+main(int argc, char *argv[]) {
+  int tickets = -1;
+  if (argc > 1) {
+    tickets = atoi(argv[1]);
+  }
+
+
+  settickets(tickets);
+  int dummy=0;
+
+  for (int i = 0; i < 5; i++) {
+    int pid = fork();
+    if (pid == 0) {
+      break;
+    }
+
+    if (pid == 0) {
+      for (int j = 0; j < NUM_ITERATION; j++) {
+        if(NUM_ITERATION%5==0) sleep(100);
+      }
+    }
+    else
+    { 
+      for (int j = 0; j < NUM_ITERATION; j++) {
+        dummy=dummy^dummy;
+      }
+    }
+ 
+  }
+  dummy=1;
+  int ck=0;
+  if(ck)printf("%d",dummy);
+
+
+      // for (int j = 0; j < NUM_ITERATION; j++) {
+      //   dummy=dummy^dummy;
+      // }
+
+  exit(0);
+}
diff --git a/user/history.c b/user/history.c
new file mode 100644
index 0000000..e0e0f2d
--- /dev/null
+++ b/user/history.c
@@ -0,0 +1,37 @@
+#include "kernel/types.h"
+#include "kernel/fcntl.h"
+#include "user/user.h"
+#include "kernel/syscall_stat.h"
+
+
+
+void print_syscall_stat(int sysno, struct syscall_stat *stat) {
+    printf("%d: syscall: %s, #: %d, time: %d\n",
+           sysno,
+           stat->syscall_name,
+           stat->count,
+           stat->accum_time);
+}
+
+
+int main(int argc, char *argv[]) {
+  if(argc == 1){
+    for(int i = 1; i <= NSYSCALLS; i++){
+      struct syscall_stat stat;
+      if(history(i, &stat) == 0){
+        print_syscall_stat(i, &stat);
+      } else {
+        printf("history fetching failed");
+      }
+    }
+  }else{
+    int sysno = atoi(argv[1]);
+    struct syscall_stat stat;
+    if(history(sysno, &stat) == 0){
+      print_syscall_stat(sysno, &stat);
+    }else{
+      printf("history fetching failed");
+    }
+  }
+  exit(0);
+}
diff --git a/user/testprocinfo.c b/user/testprocinfo.c
new file mode 100644
index 0000000..fa82bd8
--- /dev/null
+++ b/user/testprocinfo.c
@@ -0,0 +1,31 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/pstat.h"
+
+
+
+int main() {
+    struct pstat ps;
+
+    if (getpinfo(&ps) < 0) {
+        printf("getpinfo syscall failed\n");
+        exit(1);
+    }
+ 
+    printf("PID\tInUse\tQueue\tOrigTickets\tCurrTickets\tTimeSlices\n");
+    for (int i = 0; i < 64; i++) { 
+        if (ps.pid[i]!=0) {
+            printf("%d\t%d\t%d\t%d\t\t%d\t\t%d\n",
+                   ps.pid[i],
+                   ps.inuse[i],
+                   ps.inQ[i],
+                   ps.tickets_original[i],
+                   ps.tickets_current[i],
+                   ps.time_slices[i]);
+        }
+    }
+    
+
+    exit(0);
+}
diff --git a/user/user.h b/user/user.h
index f16fe27..df5509d 100644
--- a/user/user.h
+++ b/user/user.h
@@ -1,5 +1,8 @@
+struct syscall_stat;
+struct pstat;
 struct stat;
 
+
 // system calls
 int fork(void);
 int exit(int) __attribute__((noreturn));
@@ -22,6 +25,9 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
+int history(int ,struct syscall_stat *);
+int settickets(int);
+int getpinfo(struct pstat*);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..397fec1 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,6 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("history");
+entry("settickets");
+entry("getpinfo");
\ No newline at end of file
